File: kvcompat/src/main/java/io/github/porum/kvcompat/utils/AppUtils.java
package io.github.porum.kvcompat.utils;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Process;
import android.text.TextUtils;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.List;

import io.github.porum.kvcompat.logger.LogUtils;

public class AppUtils {
  private static final String TAG = "AppUtils";

  private static volatile String currentProcessName = "";

  public static boolean isMainProcess(Context context) {
    return TextUtils.equals(context.getPackageName(), getCurrentProcessName(context));
  }

  public static String getCurrentProcessName(Context context) {
    if (TextUtils.isEmpty(currentProcessName)) {
      currentProcessName = getProcessNameByPid(context, Process.myPid());
    }
    return currentProcessName;
  }

  private static String getProcessNameByPid(Context context, int pid) {
    try {
      if (context != null) {
        ActivityManager am = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        if (am != null) {
          List<ActivityManager.RunningAppProcessInfo> list = am.getRunningAppProcesses();
          if (list != null && !list.isEmpty()) {
            for (ActivityManager.RunningAppProcessInfo info : list) {
              if (info.pid == pid) {
                return info.processName;
              }
            }
          }
        }
      }
    } catch (Throwable th) {
      LogUtils.w(TAG, "getProcessNameByPid error.", th);
    }

    return getProcessNameByCmdline(pid);
  }

  private static String getProcessNameByCmdline(int pid) {
    try (BufferedReader br = new BufferedReader(new FileReader("/proc/" + pid + "/cmdline"))) {
      String processName = br.readLine();
      if (!TextUtils.isEmpty(processName)) {
        processName = processName.trim().intern();
      }
      return processName;
    } catch (Throwable tr) {
      LogUtils.w(TAG, "getProcessNameByCmdline error.", tr);
    }

    return null;
  }
}

File: kvcompat/src/main/java/io/github/porum/kvcompat/utils/AppUtils.kt
package io.github.porum.kvcompat.utils

import android.app.ActivityManager
import android.content.Context
import android.os.Process
import java.io.File

internal object AppUtils {

  @Volatile
  private var currentProcessName: String? = null

  @JvmStatic
  fun isMainProcess(context: Context) = context.packageName == getCurrentProcessName(context)

  @JvmStatic
  fun getCurrentProcessName(context: Context): String? {
    return currentProcessName ?: (getProcessNameByRunningInfo(context, Process.myPid())
      ?: getProcessNameByCmdline(Process.myPid()))?.also {
      currentProcessName = it
    }
  }

  private fun getProcessNameByRunningInfo(context: Context, pid: Int): String? {
    val am = context.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager ?: return null
    return am.runningAppProcesses.firstOrNull { it.pid == pid }?.processName
  }

  @Suppress("SpellCheckingInspection")
  private fun getProcessNameByCmdline(pid: Int): String? {
    return File("/proc/$pid/cmdline").bufferedReader().use {
      readLine()?.trim()?.intern()
    }
  }
}